### PR TITLE
few properties were not getting updated via rules on form initialization

### PR DESCRIPTION
**Pull request by barshat7**
## Description

Some Properties like 'required' 'description' and 'label' were not getting updated when set via rule 'When Form is Initialized'.
Further, the 'description' was never getting updated as if no description was authored initially then the description div was explicitly ignored in the Sightly. As a fix, included the long description div (which is always hidden by default until question mark icon is clicked), and marked the questionMark div as hidden unless we see that the description is present.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes and the overall coverage did not decrease.
- [ ] All unit tests pass on CircleCi.
- [ ] I ran all tests locally and they pass.

### CODE REVIEW

File: ui.tests/test-module/libs/support/index.js [hunk 1]
```diff
@@ -111,6 +111,11 @@ Cypress.on('uncaught:exception', (err, runnable) => {
         return false;
     }
 
+    // this error can come if properties are set via rule editor on form initialization (Since field are yet not initialized completely)
+    if (err.message.includes('some error occurred while triggering event on guideBridge')) {
```
Comment (rismehta) [inline ui.tests/test-module/libs/support/index.js:115]: The test case should fail, why is this part of whitelist ? Please remove this

File: ui.frontend/src/view/FormFieldBase.js [hunk 2]
```diff
@@ -332,25 +335,29 @@ class FormFieldBase extends FormField {
...
+                panelName,
+                fieldQualifiedName,
+                ...(originalEventPayload?.prevValue !== undefined ? { prevText: originalEventPayload?.prevValue } : {}),
+                ...(originalEventPayload?.currentValue !== undefined ? { newText: originalEventPayload?.currentValue } : {}),
+                ...((typeof originalEventPayload === 'object' && originalEventPayload !== null) ? originalEventPayload : {})
+            };
+            const formContainerPath = this.formContainer.getPath();
+            window.guideBridge.trigger(eventType, eventPayload, formContainerPath);
+        } catch (e) {
+            console.log('some error occurred while triggering event on guideBridge', e);
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:359]: Why did you add a try catch here, ideally the code should break. Can you describe the use-case

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

File: ui.tests/test-module/specs/tabsontop/tabsontop.runtime.spec.js [hunk 3]
```diff
@@ -263,13 +263,13 @@ describe("setFocus on the first invalid field on submission", () => {
     });
 
     it("check if invalid field gets focus on submit button click", () => {
-           cy.get('button').should('be.visible').click().then(() => {
+           cy.get('button').last().should('be.visible').click().then(() => {
```
Comment (rismehta) [inline ui.tests/test-module/specs/tabsontop/tabsontop.runtime.spec.js:266]: Please don't change anything in an existing test case, unless there is backward incompatible change

Comment (barshat7) [reply ui.tests/test-module/specs/tabsontop/tabsontop.runtime.spec.js:266]: Previously we did not include the questionMark div if long description was not set, but since the setProperty can be called at runtime and even if long description was not set we need to set it in runtime, then we need to have this div. As per this change https://github.com/adobe/aem-core-forms-components/pull/1274/files#:~:text=%3Cbutton%20class%3D%22%24%7BbemBlock%7D__questionmark%22%20style%3D%22%24%7BlongDescription%20%3F%20%27%27%20%3A%20%27display%3Anone%3B%27%7D%20title%3D%22Help%20text%22%3E

Now in the test cases the selectors were wrong as it was assuming that there would only be one button in the form, so I had to update the selectors. 
Not sure if this would be backward incompatible though

File: ui.frontend/src/view/FormFieldBase.js [hunk 4]
```diff
@@ -602,7 +608,17 @@ class FormFieldBase extends FormField {
...
-            this.description.querySelector("p").innerHTML = description;
+            if (description) {
+                if (!this.description.querySelector("p")) {
+                    // If the description is updated via rule then it might not have <p> tags
+                    this.description.appendChild(document.createElement('p'));
+                }
+                this.description.querySelector("p").innerHTML = description;
+                const questionMarkDiv = this.getQuestionMarkDiv();
+                 if (questionMarkDiv) {
+                     questionMarkDiv.style.display = "block";
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:619]: We should not be updating styles via code, this might override the theme, and might conflict. What is the use-case to override style

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/questionMark.html [hunk 5]
```diff
@@ -1,4 +1,4 @@
 <template data-sly-template.questionMark="${@ componentId, longDescription, bemBlock}">
-  <button class="${bemBlock}__questionmark" data-sly-test="${longDescription}" title="Help text">
+  <button class="${bemBlock}__questionmark" style="${longDescription ? '' : 'display:none;'} title="Help text">
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/questionMark.html:2]: styles need to come via theme, code should not add inline style, it would override other styles provided by customer. This looks incorrect to me

File: ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/longDescription.html [hunk 6]
```diff
@@ -1,5 +1,5 @@
 <template data-sly-template.longDescription="${@ componentId, longDescription, bemBlock}">
-    <div class="${bemBlock}__longdescription" id="${componentId}__longdescription" data-sly-test="${longDescription}" aria-live="polite">
-      ${longDescription @context='html'}
-    </div>
-</template>
\ No newline at end of file
+  <div class="${bemBlock}__longdescription" id="${componentId}__longdescription" style="${wcmmode.edit && longDescription ? '' : 'display:none;'} aria-live="polite">
```
Comment (rismehta) [inline ui.af.apps/src/main/content/jcr_root/apps/core/fd/components/af-commons/v1/fieldTemplates/longDescription.html:2]: Why do we need inline style

File: ui.frontend/src/view/FormFieldBase.js (hunk 4 above)
Comment (barshat7) [reply ui.frontend/src/view/FormFieldBase.js:619]: The requirement is that if long description is not set during authoring then the questionMark div and description div should be present in DOM but hidden. And if by any rule the description is set in runtime, then they should be visible. I guess we could do that by adding 'data-cmp-visible' attribute instead

File: ui.tests/test-module/specs/tabsontop/tabsontop.runtime.spec.js (hunk 3 above)
Comment (rismehta) [reply ui.tests/test-module/specs/tabsontop/tabsontop.runtime.spec.js:266]: But this would pollute the HTML with unncessary markup for each field, thus can impact HTML parsing even for use-cases where description is not set at runtime, which is incorrect.

Ideally, at runtime, we should generate the markup if description is set at runtime.

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

File: ui.frontend/src/view/FormFieldBase.js (hunk 4 above)
Comment (rismehta) [reply ui.frontend/src/view/FormFieldBase.js:619]: yes, please check this comment on the overall design, https://github.com/adobe/aem-core-forms-components/pull/1274#discussion_r1654523515

**Comment by rismehta** [comment]
Also, do we have a customer who has asked for this change ? Can you link the JIRA in the PR

File: ui.frontend/src/view/FormFieldBase.js [hunk 7]
```diff
@@ -598,16 +604,45 @@ class FormFieldBase extends FormField {
...
+        if (descriptionText !== undefined) {
+            if (this.getDescription()) {
+                if (!this.getDescription().querySelector("p")) {
+                    // If the description is updated via rule then it might not have <p> tags
+                    this.getDescription().appendChild(document.createElement('p'));
+                }
+            } else {
+                this.#addDescriptionInRuntime();
+            }
+            this.getDescription().querySelector("p").innerHTML = descriptionText;
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:619]: You would need to sanitize this HTML, we already do this for enumNames,

```
window.DOMPurify ?  window.DOMPurify.sanitize(descriptionText) : descriptionText
```

File: ui.frontend/src/view/FormFieldBase.js [hunk 8]
```diff
@@ -598,16 +604,45 @@ class FormFieldBase extends FormField {
...
-     */
-    updateDescription(description) {
-        if (this.description) {
-            this.description.querySelector("p").innerHTML = description;
-        } else {
-            //TODO: handle the case when description is not present initially.
+     * @param {string} descriptionText - The description.
+     */
+    updateDescription(descriptionText) {
+        if (descriptionText !== undefined) {
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:610]:
```  
  if (typeof descriptionText !== 'undefined') {
        const descriptionElement = this.getDescription();
        if (descriptionElement) {
            let pElement = descriptionElement.querySelector("p");
            if (!pElement) {
                // If the description is updated via rule then it might not have <p> tags
                pElement = document.createElement('p');
                descriptionElement.appendChild(pElement);
            }
            pElement.textContent = descriptionText;
        } else {
            this.#addDescriptionInRuntime(descriptionText);
        }
    }
    ```

File: ui.frontend/src/view/FormFieldBase.js [hunk 9]
```diff
@@ -598,16 +604,45 @@ class FormFieldBase extends FormField {
...
+                }
+            } else {
+                this.#addDescriptionInRuntime();
+            }
+            this.getDescription().querySelector("p").innerHTML = descriptionText;
         }
     }
 
+    #addDescriptionInRuntime() {
+        console.log('adding description in runtime')
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:624]: Remove the console log

File: ui.frontend/src/view/FormFieldBase.js [hunk 10]
```diff
@@ -598,16 +604,45 @@ class FormFieldBase extends FormField {
...
+        // add question mark icon
+        const bemClass = Array.from(this.element.classList).filter(bemClass => !bemClass.includes('--'))[0];
```
Comment (rismehta) [inline ui.frontend/src/view/FormFieldBase.js:626]: lets make the code NPE safe

```
 // add question mark icon
    const bemClass = Array.from(this.element.classList).find(bemClass => !bemClass.includes('--'));
    const labelContainer = this.element.querySelector(`.${bemClass}__label-container`);
    if (labelContainer) {
        const qmButton = document.createElement('button');
        qmButton.className = `${bemClass}__questionmark`;
        qmButton.title = 'Help text';
        labelContainer.appendChild(qmButton);
    } else {
        console.error('Label container not found');
        return;
    }

    // add description div
    const descriptionDiv = document.createElement('div');
    descriptionDiv.className = `${bemClass}__longdescription`;
    descriptionDiv.id = `${this.getId()}__longdescription`;
    descriptionDiv.setAttribute('aria-live', 'polite');
    descriptionDiv.setAttribute('data-cmp-visible', 'false');
    
    const pElement = document.createElement('p');
    if (typeof descriptionText !== 'undefined') {
        pElement.textContent = descriptionText;
    }
    descriptionDiv.appendChild(pElement);

    const errorDiv = this.getErrorDiv();
    if (errorDiv) {
        this.element.insertBefore(descriptionDiv, errorDiv);
    } else {
        console.error('Error div not found');
        return;
    }
```

**Review by rismehta — CHANGES_REQUESTED** [review]
check comments

**Review by rismehta — APPROVED** [review]
